src/node/ledger_secrets.h

```diff
@@ -54,28 +54,38 @@ namespace ccf
 
     SpinLock lock;
     LedgerSecretsMap ledger_secrets;
-    std::optional<LedgerSecretsMap::iterator> commit_key_it = std::nullopt;
 
-    LedgerSecretsMap::iterator get_encryption_key_it(kv::Version version)
+    LedgerSecret get_secret_for_version(kv::Version version)
     {
-      // Encryption key for a given version is the one with the highest version
-      // that is lower than the given version (e.g. if encryption_keys contains
-      // two keys for version 0 and 10 then the key associated with version 0
-      // is used for version [0..9] and version 10 for versions 10+)
+      // The ledger secret used to encrypt/decrypt a transaction at a given
+      // version is the one with the highest version that is lower than the
+      // given version (e.g. if ledger_secrets contains two keys for version 0
+      // and 10 then the key associated with version 0 is used for version
+      // [0..9] and version 10 for versions 10+)
 
-      auto search_begin = commit_key_it.value_or(ledger_secrets.begin());
+      if (ledger_secrets.empty())
+      {
+        throw std::logic_error("Ledger secrets map is empty");
+      }
+
+      if (version >= ledger_secrets.rbegin()->first)
+      {
+        // Hot path for encrypting/decrypting transactions since the latest
+        // rekey
+        return ledger_secrets.rbegin()->second;
+      }
 
       auto search = std::upper_bound(
-        search_begin, ledger_secrets.end(), version, [](auto a, const auto& b) {
-          return b.first > a;
-        });
-      if (search == search_begin)
+        ledger_secrets.begin(),
+        ledger_secrets.end(),
+        version,
+        [](auto a, const auto& b) { return b.first > a; });
+      if (search == ledger_secrets.begin())
       {
-        throw std::logic_error(fmt::format(
-          "kv::TxEncryptor: could not find ledger encryption key for seqno {}",
-          version));
+        throw std::logic_error(
+          fmt::format("Could not find ledger secret for seqno {}", version));
       }
-      return --search;
+      return (--search)->second;
```

Comment:
```suggestion
      return std::prev(search)->second;
```

I normally prefer `++` and `--` for iterator manipulation, but for some reason the additional lookup here makes it look like dangerous pointer chasing, so `std::prev` would make my eyes happier. Very minor!